### PR TITLE
 Authentication-Spring-Security

### PR DESCRIPTION
**Pull request by Peachex**
https://github.com/mjc-school/MJC-School/blob/master/stage%20%233/java/module%20%234.%20Authentication%20%26%20Spring%20Security/authentication_and_spring_security_task.md

### CODE REVIEW

File: repository/src/main/java/com/epam/esm/dao/creator/criteria/search/FullMatchSearchCertificateCriteria.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    public FullMatchSearchCertificateCriteria(List<Tag> tags) {
+        super(tags);
+    }
+
+    @Override
+    public void acceptCriteria(CriteriaQuery<GiftCertificate> criteriaQuery, CriteriaBuilder builder,
+                               Root<GiftCertificate> root) {
+        List<Tag> searchedTags = getTags();
+        if (!CollectionUtils.isEmpty(searchedTags)) {
```
Comment (AliakseiYermachonak) [inline repository/src/main/java/com/epam/esm/dao/creator/criteria/search/FullMatchSearchCertificateCriteria.java:31]: doesn't it have smth like 'isNotEmpty'?

File: repository/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+public class GiftCertificateDaoImpl implements GiftCertificateDao<GiftCertificate> {
+    private final QueryCreator<GiftCertificate> criteriaCreator;
+    private EntityManager manager;
+
+    @Autowired
+    public GiftCertificateDaoImpl(QueryCreator<GiftCertificate> criteriaCreator) {
+        this.criteriaCreator = criteriaCreator;
+    }
+
+    @PersistenceContext
```
Comment (AliakseiYermachonak) [inline repository/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java:33]: interesting

File: repository/src/main/java/com/epam/esm/dao/creator/criteria/search/FullMatchSearchCertificateCriteria.java (hunk 1 above)
Comment (Peachex) [reply repository/src/main/java/com/epam/esm/dao/creator/criteria/search/FullMatchSearchCertificateCriteria.java:31]: Yes, I have already found this method.

File: repository/src/main/java/com/epam/esm/dto/Order.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Column(name = "timestamp")
+    private LocalDateTime timestamp;
+
+    @ManyToOne(fetch = FetchType.EAGER)
+    @JoinColumn(
+            name = "gift_certificate_id_fk"
+    )
+    private GiftCertificate giftCertificate;
+
```
Comment (AliakseiYermachonak) [inline repository/src/main/java/com/epam/esm/dto/Order.java:53]: image

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                if (!existingTags.isEmpty()) {
+                    existingTags.stream().filter(t -> !t.isAvailable()).peek(t -> t.setAvailable(true))
+                            .forEach(t -> tagService.updateAvailability(String.valueOf(t.getId()), true));
+                }
+                Set<Tag> newTags = new HashSet<>(CollectionUtils.removeAll(giftCertificate.getTags(), existingTags));
+                newTags.forEach(t -> t.setAvailable(true));
+                giftCertificate.setTags(newTags);
+
+                id = dao.insert(giftCertificate);
+                GiftCertificate certificateWithAllTags = dao.findById(id).get();
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:72]: Nope

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    @Override
+    public boolean update(String id, GiftCertificate newCertificate) {
+        try {
+            GiftCertificate oldCertificate = dao.findById(Long.parseLong(id)).orElseThrow(() ->
+                    new ResourceNotFoundException(ErrorCode.GIFT_CERTIFICATE,
+                            ErrorName.RESOURCE_NOT_FOUND, id));
+
+            if (updateCertificateFields(oldCertificate, newCertificate)) {
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:115]: update is more an action than asking for the boolean result

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService<GiftCertificate> {
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:34]: methods are rather complex in the class

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Override
+    public Tag findMostUsedTagOfUserWithHighestCostOfAllOrders(String userId) {
+        return dao.findMostUsedTagOfUserWithHighestCostOfAllOrders(userService.findById(userId).getId()).orElseThrow(
+                () -> new ResourceNotFoundException(ErrorCode.TAG, ErrorName.RESOURCE_NOT_FOUND, userId));
+    }
+
+    @Override
+    public List<Tag> findAll(int page, int elements) {
+        if (page < 1 || elements < 1) {
```
Comment (AliakseiYermachonak) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java:87]: is there some other way to validate the parameters in the methods?

File: web/src/main/java/com/epam/esm/jwt/JwtFilter.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            httpResponse.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+
+            HttpServletRequest httpRequest = (HttpServletRequest) request;
+            ExceptionResponse exceptionResponse = new ExceptionResponse(e.getLocalizedMessage(MessageLocale.defineLocale(
+                    httpRequest.getHeader(HeaderName.LOCALE))), e.getErrorCode());
+            exceptionResponse.setErrorCode(HttpStatus.UNAUTHORIZED.value() + e.getErrorCode());
+
+            ResponseEntity<ExceptionResponse> responseEntity = new ResponseEntity<>(exceptionResponse, HttpStatus.UNAUTHORIZED);
+            httpResponse.getOutputStream().println(jsonConverter.convert(responseEntity));
+            return;
```
Comment (AliakseiYermachonak) [inline web/src/main/java/com/epam/esm/jwt/JwtFilter.java:58]: hmm

**Pull request merged**